authority/tls.go

```diff
@@ -251,6 +252,52 @@ func (a *Authority) Renew(ocx *x509.Certificate) (*x509.Certificate, *x509.Certi
 	return serverCert, caCert, nil
 }
 
+// RevocationTypeSelector is a map representing the selected types of
+type RevocationTypeSelector struct {
+	All, Passive, CRL, OCSP bool
+}
+
+// Revoke revokes a certificate.
+//
+// NOTE: Only supports passive revocation - prevent existing certificates from
+// being renewed.
+//
+// TODO: Add OCSP and CRL support.
+func (a *Authority) Revoke(rts RevocationTypeSelector, serial string, provisionerID string, reasonCode int) error {
+	var errContext = context{
+		"serialNumber":  serial,
+		"provisionerID": provisionerID,
+		"reasonCode":    reasonCode,
+	}
+	if rts.All || rts.Passive {
+		if a.db == nil {
+			return &apiError{errors.New("no persistence layer configured"),
+				http.StatusNotImplemented, errContext}
+		}
+
+		rci := &db.RevokedCertificateInfo{
+			Serial:        serial,
+			ProvisionerID: provisionerID,
+			ReasonCode:    reasonCode,
+			RevokedAt:     time.Now().UTC(),
+		}
+
+		if err := a.db.Revoke(rci); err != nil {
+			return &apiError{err, http.StatusInternalServerError, errContext}
+		}
+	}
+	if rts.All || rts.CRL {
+		return &apiError{errors.New("revoke CRL unimplemented"), http.StatusNotImplemented,
+			errContext}
+	}
+	if rts.All || rts.OCSP {
+		return &apiError{errors.New("revoke OCSP unimplemented"), http.StatusNotImplemented,
+			errContext}
+	}
```

Comment:
If `rts.All` is `true` it will create a passive revocation and always return an error.

Reply:
lol, good call.